myhoard/basebackup_restore_operation.py

```diff
@@ -313,6 +322,10 @@ def _process_xbstream_output_line(self, line, _stream_name):
         if not line:
             return
 
+        if self._process_xtrabackup_output_line_no_space_left(line):
+            self.log.error("xbstream: %r", line)
+            raise DiskFullError("No space left on device. Cannot complete xbstream-extract!")
```

Comment:
These blocks are repetitive. The `_process_xtrabackup_output_line_no_space_left` function already logs the line so logging it here is not necessary. Also error could be raised from the function so that the callers don't need to do anything extra. (In practice only the `xbstream` phase is in practice potentially running out of disk so doing this for others is not that relevant but might be a good idea just in case.)

Reply:
I think `_process_xtrabackup_output_line_no_space_left` should not log. It's a predicate function, it should not have logging side-effects. The reason why it logs, is because I reused the existing patterns.

The repetitive blocks then "become" correct, as `_process_xtrabackup_output_line_no_space_left` should not have more responsibilities.
It would also be possible to have yet another method, which would be something like:
```python
    def _raise_if_disk_full(self, line: str, bin_name: str, caller_id: str) -> None:
        if self._process_xtrabackup_output_line_no_space_left(line):
            self.log.error("%s: %r", bin_name, line)
            raise DiskFullError("No space left on device. Cannot complete %s!", caller_id)
```

But it really looks like a "util" function, to force DRY-ness. I'm not super excited about that, and would prefer having these 3 blocks. Especially as they could have different logic and should not depend on the other binary's stdout/stderr.

Reply:
I would go with the extra function to avoid duplication but I can anyway leave this pr for the team actually working on this component so no need to mind me. I just had a quick glance as I was curious to see if we can easily detect the "out of disk space" situation (seems like we can).

Reply:
I tend to agree with @rikonen here, a separate function makes sense as the log-message is basically the same.

Reply:
Changed in new version. 🚀